src/libstd/thread/mod.rs

```diff
@@ -1192,7 +1194,7 @@ impl<T> JoinInner<T> {
 ///     });
 /// });
 ///
-/// let _ = original_thread.join();
+/// original_thread.join().expect("Unable to join thread");
```

Comment:
Join will return an error if the other thread panicked, so this error message may not be the most accurate.

Reply:
Is it then correct to say that there two failure cases that should be discussed in each case?

1. The thread panicked.
1. We couldn't `join` (whatever underlying error).

---

The existing docs for `JoinHandle::join` say:

```
join_handle.join().expect("Couldn't join on the associated thread");
```

Would you like to see both of the above cases mentioned in every expect? Just in the example for `JoinHandle::join` and hope people read it? Something else? 

I always regret adding `unwrap` to examples because that teaches people to not have useful tracer messages, but saying nothing useful avoids saying something wrong.

Reply:
I would think the JonHandle docs should change as well.

Reply:
> I would think the JoinHandle docs should change as well.

To be clear, you'd like all the examples to have something like this?

```rust
.join().expect("Unable to join thread or thread panicked")
```

Reply:
Potentially, though the contexts in which the error was due to a join error are really limited.

Reply:
This is personally why I use `unwrap` over `expect`; it's tough to get a good message.

@sfackler , if you could be more concrete here, that'd be great. I'm happy with whatever, but right now, it's unclear what exactly @shepmaster needs to do for this to land.

Reply:
> This is personally why I use `unwrap` over `expect`; it's tough to get a good message.

Wait what?! 😨 It's always better to have a message than just nothing!

Reply:
@shepmaster Docs indicate that the only error that .join() can return is if the thread has panicked

Reply:
@bluss good point; checking [the implementation](https://github.com/rust-lang/rust/blob/1.20.0/src/libstd/thread/mod.rs#L1086-L1093) seems to bear that out:

```rust
impl<T> JoinInner<T> {
    fn join(&mut self) -> Result<T> {
        self.native.take().unwrap().join();
        unsafe {
            (*self.packet.0.get()).take().unwrap()
        }
    }
}
``` 

The actual join has no potential `Result`. Based on that, it seems that the correct course of action is that the messages should say something akin to `expect("the thread being joined has panicked")`.